cranelift-codegen/src/regalloc/liveness.rs

```diff
@@ -448,6 +448,43 @@ impl Liveness {
                         lr.affinity.merge(constraint, &reginfo);
                     }
                 }
+
+                if let Some(sig) = func.dfg.call_signature(inst) {
+                    let sig = &func.dfg.signatures[sig];
+                    let args = func.dfg.inst_variable_args(inst);
+                    if sig.params.len() != args.len() {
+                        panic!(
+                            "Assumption violated: args count == sig.params count ({} != {})",
+                            args.len(),
+                            sig.params.len()
+                        );
+                    }
+                    for (param, &arg) in sig.params.iter().zip(args.iter()) {
+                        // Safety: lr was ensured to exist through get_or_create in the above
+                        // iteration of all inst args (of which inst_variable_args is a subset)
+                        let lr = self.ranges.get_mut(arg).unwrap();
+                        lr.affinity = match (lr.affinity, Affinity::abi(param)) {
+                            (Affinity::Unassigned, abi) => abi,
+                            (Affinity::RegClass(rci), Affinity::RegUnit(unit)) => {
+                                if reginfo.rc(rci).contains(unit) {
+                                    Affinity::RegUnit(unit)
+                                } else {
+                                    Affinity::RegClass(rci)
+                                }
+                            }
+                            (Affinity::RegClass(lr_rci), Affinity::RegClass(abi_rci)) => {
+                                if let Some(subclass) =
+                                    reginfo.rc(lr_rci).intersect_index(reginfo.rc(abi_rci))
+                                {
+                                    Affinity::RegClass(subclass)
+                                } else {
+                                    Affinity::RegClass(lr_rci)
+                                }
+                            }
+                            (lr_affinity, _) => lr_affinity,
+                        };
```

Comment:
This `match` expression looks like a good candidate to be refactored out into a helper function.

Reply:
:+1: this is now `Affinity::intersect`